File: scala/pal/src/main/scala/pal/PhasePeg.scala
package pal

import java.util.concurrent.atomic.AtomicInteger
import scala.collection.immutable.VectorMap
import scala.collection.mutable

/** Experimental inverse expansion for the plain PEG subset emitted by tm2peg.
  *
  * `inverseRepeat(G, k)` recognizes {w | G recognizes each character of w k times}.
  * The virtual offset inside a repeated character is a finite grammar parameter,
  * not runtime state. `E_i_j` succeeds exactly when E returns in phase j from i.
  * An ordered-choice fallback is guarded by failure of E in *every* return phase.
  *
  * This is an output-boundary construction, not a real-time scheduling proof.
  * Input is BMP scalar text. The parser deliberately supports only double-quoted
  * literals, '.', names, sequence, choice, predicates, grouping and '*'. Source
  * grammars must terminate (in particular no nullable repetition/left recursion).
  *
  * Port of `phase_peg.py`.
  */
object PhasePeg {

  /** The Python `EMPTY` singleton. */
  val EMPTY: PegAst = PegAst.Empty

  private val NAME = "[A-Za-z_][A-Za-z_0-9]*".r
  private val TOKEN = "[A-Za-z_][A-Za-z_0-9]*|[=;()/!&.*]".r

  private[pal] def isName(text: String): Boolean = NAME.matches(text)

  private def isBmpScalar(c: Char): Boolean = !Character.isSurrogate(c)

  /** Python `json.dumps(c, ensure_ascii=False)` for one character, with backspace spelled ``. */
  def literal(c: Char): String = {
    val escaped = c match {
      case '\b' => "\\u0008"
      case '"' => "\\\""
      case '\\' => "\\\\"
      case '\n' => "\\n"
      case '\r' => "\\r"
      case '\t' => "\\t"
      case '\f' => "\\f"
      case control if control < 0x20 => f"\\u${control.toInt}%04x"
      case other => other.toString
    }
    "\"" + escaped + "\""
  }

  private[pal] def requireBmp(text: String, message: String): Unit = {
    if (text.exists(c => !isBmpScalar(c))) { throw new IllegalArgumentException(message) }
  }

  // ---------------------------------------------------------------------
  // Tokenizer (the first half of Python `Grammar.__init__`).
  // ---------------------------------------------------------------------

  private[pal] sealed trait Token
  private[pal] final case class Literal(value: String) extends Token
  private[pal] final case class Punct(text: String) extends Token

  private[pal] def tokenize(source: String): Vector[Token] = {
    val tokens = Vector.newBuilder[Token]
    val matcher = TOKEN.pattern.matcher(source)
    var cursor = 0
    while (cursor < source.length) {
      val c = source.charAt(cursor)
      if (Character.isWhitespace(c) || Character.isSpaceChar(c)) {
        cursor += 1
      } else if (c == '"') {
        val (value, end) = JsonString.decode(source, cursor)
        requireBmp(value, "BMP scalar literals required")
        tokens += Literal(value)
        cursor = end
      } else {
        matcher.region(cursor, source.length)
        if (!matcher.lookingAt()) { throw new IllegalArgumentException(s"unsupported PEG syntax at $cursor") }
        tokens += Punct(matcher.group())
        cursor = matcher.end()
      }
    }
    tokens.result()
  }

  /** Decodes one JSON string literal starting at `start` (Python `json.JSONDecoder.raw_decode`). */
  private object JsonString {
    def decode(source: String, start: Int): (String, Int) = {
      val out = new StringBuilder
      var cursor = start + 1
      var closed = false
      while (!closed) {
        if (cursor >= source.length) { throw new IllegalArgumentException("unterminated string literal") }
        val c = source.charAt(cursor)
        if (c == '"') {
          closed = true
          cursor += 1
        } else if (c == '\\') {
          cursor = escape(source, cursor, out)
        } else if (c < 0x20) {
          throw new IllegalArgumentException(s"invalid control character at $cursor")
        } else {
          out += c
          cursor += 1
        }
      }
      (out.toString, cursor)
    }

    private def escape(source: String, cursor: Int, out: StringBuilder): Int = {
      if (cursor + 1 >= source.length) { throw new IllegalArgumentException("unterminated string literal") }
      source.charAt(cursor + 1) match {
        case '"' => out += '"'; cursor + 2
        case '\\' => out += '\\'; cursor + 2
        case '/' => out += '/'; cursor + 2
        case 'b' => out += '\b'; cursor + 2
        case 'f' => out += '\f'; cursor + 2
        case 'n' => out += '\n'; cursor + 2
        case 'r' => out += '\r'; cursor + 2
        case 't' => out += '\t'; cursor + 2
        case 'u' =>
          if (cursor + 6 > source.length) { throw new IllegalArgumentException("invalid \\u escape") }
          val hex = source.substring(cursor + 2, cursor + 6)
          if (!hex.forall(h => Character.digit(h, 16) >= 0)) { throw new IllegalArgumentException("invalid \\u escape") }
          out += Integer.parseInt(hex, 16).toChar
          cursor + 6
        case other => throw new IllegalArgumentException(s"invalid escape \\$other")
      }
    }
  }

  // ---------------------------------------------------------------------
  // Inverse fixed-width expansion.
  // ---------------------------------------------------------------------

  /** Phase-return relation of every reachable node: `(i, j)` when the node can start in phase i and return in phase j. */
  private final class PhaseReturns(grammar: Grammar, width: Int) {
    import PegAst.*

    val nodes: Vector[PegAst] = {
      val ids = mutable.LinkedHashSet.empty[PegAst]
      def register(expr: PegAst): Unit = {
        if (ids.add(expr)) {
          expr match {
            case Ref(name) => register(grammar.rules(name))
            case Sequence(first, second) => register(first); register(second)
            case Choice(first, second) => register(first); register(second)
            case NotPredicate(inner) => register(inner)
            case AndPredicate(inner) => register(inner)
            case Star(inner) => register(inner)
            case Empty | AnyChar | Terminal(_) => ()
          }
        }
      }
      register(grammar.rules(grammar.start))
      ids.toVector
    }

    val ids: Map[PegAst, Int] = nodes.zipWithIndex.toMap

    private val diagonal: Set[(Int, Int)] = (0 until width).map(i => (i, i)).toSet

    val returns: Map[PegAst, mutable.Set[(Int, Int)]] = {
      val table = nodes.map(expr => expr -> mutable.HashSet.empty[(Int, Int)]).toMap
      var changed = true
      while (changed) {
        changed = false
        for (expr <- nodes) {
          val possible = possibleReturns(expr, table)
          if (!possible.subsetOf(table(expr))) {
            table(expr) ++= possible
            changed = true
          }
        }
      }
      table
    }

    private def compose(left: collection.Set[(Int, Int)], right: collection.Set[(Int, Int)]): Set[(Int, Int)] = {
      (for ((i, k) <- left; (middle, j) <- right if middle == k) yield (i, j)).toSet
    }

    private def possibleReturns(expr: PegAst, table: Map[PegAst, mutable.Set[(Int, Int)]]): Set[(Int, Int)] = {
      expr match {
        case Empty | NotPredicate(_) => diagonal
        case Terminal(_) | AnyChar => (0 until width).map(i => (i, (i + 1) % width)).toSet
        case Ref(name) => table(grammar.rules(name)).toSet
        case Choice(first, second) => table(first).toSet | table(second).toSet
        case Sequence(first, second) => compose(table(first), table(second))
        case AndPredicate(inner) => table(inner).map { case (i, _) => (i, i) }.toSet
        case Star(inner) => diagonal | compose(table(inner), table(expr))
      }
    }

    def can(expr: PegAst, i: Int, j: Int): Boolean = returns(expr).contains((i, j))

    def name(expr: PegAst, i: Int, j: Int): String = s"E_${ids(expr)}_${i}_$j"

    def success(expr: PegAst, i: Int): String = s"Y_${ids(expr)}_$i"
  }

  def inverseRepeat(source: String, width: Int, start: String = "S"): String = {
    if (width < 1) { throw new IllegalArgumentException("positive integer width required") }
    val grammar = new Grammar(source, start)
    // Prune impossible phase returns before rendering. Besides reducing size,
    // this keeps a failed E_i_i from masquerading as a nullable prefix in the
    // target grammar's conservative left-recursion check.
    val phases = new PhaseReturns(grammar, width)
    val fail = "!\"\""
    val lines = mutable.ArrayBuffer[String](s"S = ${phases.name(grammar.rules(grammar.start), 0, 0)} !.;")
    for (expr <- phases.nodes; i <- 0 until width) {
      val choices = (0 until width).filter(j => phases.can(expr, i, j)).map(j => "&" + phases.name(expr, i, j))
      lines += phases.success(expr, i) + " = " + (if (choices.isEmpty) { fail } else { choices.mkString(" / ") }) + ";"
      for (j <- 0 until width) {
        val alternatives = phaseAlternatives(grammar, phases, width, expr, i, j)
        lines += phases.name(expr, i, j) + " = " + (if (alternatives.isEmpty) { fail } else { alternatives.mkString(" / ") }) + ";"
      }
    }
    lines.mkString("", "\n", "\n")
  }

  /** The ordered alternatives of rule `E_expr_i_j`. */
  private def phaseAlternatives(grammar: Grammar, phases: PhaseReturns, width: Int, expr: PegAst, i: Int, j: Int): Seq[String] = {
    import PegAst.*
    if (!phases.can(expr, i, j)) {
      Nil
    } else {
      expr match {
        case Empty if i == j => Seq("\"\"")
        case Terminal(_) | AnyChar if j == (i + 1) % width =>
          val terminal = expr match {
            case Terminal(c) => literal(c)
            case _ => "."
          }
          Seq(if (i == width - 1) { terminal } else { s"&($terminal)" })
        case Ref(name) => Seq(phases.name(grammar.rules(name), i, j))
        case Sequence(first, second) =>
          (0 until width).filter(k => phases.can(first, i, k) && phases.can(second, k, j))
            .map(k => phases.name(first, i, k) + " " + phases.name(second, k, j))
        case Choice(first, second) =>
          val alternatives = mutable.ArrayBuffer.empty[String]
          if (phases.can(first, i, j)) { alternatives += phases.name(first, i, j) }
          if (phases.can(second, i, j)) {
            val otherPhase = phases.returns(first).exists { case (a, b) => a == i && b != j }
            val commitment = if (otherPhase) { "!" + phases.success(first, i) + " " } else { "" }
            alternatives += commitment + phases.name(second, i, j)
          }
          alternatives.toSeq
        case NotPredicate(inner) if i == j => Seq("!" + phases.success(inner, i))
        case AndPredicate(inner) if i == j => Seq("&" + phases.success(inner, i))
        case Star(inner) =>
          val steps = (0 until width).filter(k => phases.can(inner, i, k) && phases.can(expr, k, j))
            .map(k => phases.name(inner, i, k) + " " + phases.name(expr, k, j))
          if (i == j) { steps :+ ("!" + phases.success(inner, i)) } else { steps }
        case _ => Nil
      }
    }
  }
}

/** Parsed PEG expression. Structural equality (for `inverseRepeat`'s node
  * numbering) plus a unique `identity` (for the packrat memo, which the
  * Python original keyed by `id(expr)`).
  */
sealed abstract class PegAst extends Product with Serializable {
  val identity: Int = PegAst.counter.getAndIncrement()
}

object PegAst {
  private val counter = new AtomicInteger(0)

  case object Empty extends PegAst
  final case class Terminal(char: Char) extends PegAst
  case object AnyChar extends PegAst
  final case class Ref(name: String) extends PegAst
  final case class Sequence(first: PegAst, second: PegAst) extends PegAst
  final case class Choice(first: PegAst, second: PegAst) extends PegAst
  final case class NotPredicate(inner: PegAst) extends PegAst
  final case class AndPredicate(inner: PegAst) extends PegAst
  final case class Star(inner: PegAst) extends PegAst
}

/** Packrat memo keyed by (node identity, position). Values: -1 missing, -2 in progress, else stored. */
private[pal] trait Memo {
  def get(identity: Int, position: Int): Int
  def update(identity: Int, position: Int, value: Int): Unit
}

private[pal] final class HashMemo extends Memo {
  private val cells = mutable.HashMap.empty[(Int, Int), Int]
  def get(identity: Int, position: Int): Int = cells.getOrElse((identity, position), -1)
  def update(identity: Int, position: Int, value: Int): Unit = { cells((identity, position)) = value }
}

/** Sparse rows become integer arrays when packrat cells grow dense.
  *
  * Python stored dense rows as `array("H")` (16-bit) below 65533 input
  * characters and `array("Q")` (64-bit) otherwise; `kind` reports which.
  */
final class CompactMemo(length: Int) extends Memo {
  val width: Int = length + 1
  val kind: String = if (length < 65533) { "H" } else { "Q" }
  private val missing: Long = if (kind == "H") { 65535L } else { -1L } // "Q": 2**64 - 1 as an unsigned Long
  val threshold: Int = math.max(8, width / (if (kind == "H") { 32 } else { 8 }))

  private sealed trait Row
  private final class Sparse(val cells: mutable.HashMap[Int, Long]) extends Row
  private final class Dense(val cells: Array[Long]) extends Row

  private val rows = mutable.HashMap.empty[Int, Row]

  def get(identity: Int, position: Int): Int = get(identity, position, -1)

  def get(identity: Int, position: Int, default: Int): Int = {
    rows.get(identity) match {
      case None => default
      case Some(row) =>
        val value = row match {
          case sparse: Sparse => sparse.cells.getOrElse(position, missing)
          case dense: Dense => dense.cells(position)
        }
        if (value == missing) { -1 } else if (value == missing - 1) { -2 } else { value.toInt }
    }
  }

  def update(identity: Int, position: Int, value: Int): Unit = {
    val stored = if (value < 0) { missing + value + 1 } else { value.toLong }
    rows.get(identity) match {
      case Some(dense: Dense) => dense.cells(position) = stored
      case other =>
        val sparse = other match {
          case Some(existing: Sparse) => existing
          case _ =>
            val fresh = new Sparse(mutable.HashMap.empty)
            rows(identity) = fresh
            fresh
        }
        sparse.cells(position) = stored
        if (sparse.cells.size > threshold) {
          val dense = Array.fill(width)(missing)
          for ((index, cell) <- sparse.cells) { dense(index) = cell }
          rows(identity) = new Dense(dense)
        }
    }
  }
}

/** A plain PEG (the tm2peg output subset) with an explicit-continuation packrat evaluator.
  *
  * `rules` is insertion ordered for parsed sources; `FileGrammar` substitutes a
  * lazily loading map.
  */
class Grammar protected (val rules: collection.Map[String, PegAst], val start: String) {

  def this(source: String, start: String = "S") = this(Grammar.parseRules(source, start), start)

  def accepts(word: String, compact: Boolean = false): Boolean = parsePrefix(word, compact).contains(word.length)

  /** Return the start rule's end position, or `None` on failure.
    *
    * Pointer-like PEG rules deliberately leave a suffix. Exposing their result
    * lets construction experiments check that suffix, rather than only checking
    * whole-word acceptance.
    */
  def parsePrefix(word: String, compact: Boolean = false): Option[Int] = {
    PhasePeg.requireBmp(word, "BMP scalar input required")
    val memo: Memo = if (compact) { new CompactMemo(word.length) } else { new HashMemo }
    new PackratRun(this, word, memo).run()
  }
}

object Grammar {

  /** Parse `source` into an insertion-ordered rule table; `start` must be defined. */
  def parseRules(source: String, start: String): VectorMap[String, PegAst] = {
    val rules = new GrammarParser(PhasePeg.tokenize(source)).parseRules()
    if (rules.isEmpty) { throw new IllegalArgumentException("empty grammar") }
    if (!rules.contains(start)) { throw new IllegalArgumentException(s"unknown start rule: $start") }
    rules
  }
}

/** Recursive-descent parser over the token stream (Python `Grammar.choice/sequence/prefix`). */
private final class GrammarParser(tokens: Vector[PhasePeg.Token]) {
  import PhasePeg.{Literal, Punct, Token}
  import PegAst.*

  private var cursor = 0

  def parseRules(): VectorMap[String, PegAst] = {
    var rules = VectorMap.empty[String, PegAst]
    while (cursor < tokens.length) {
      val name = take() match {
        case Punct(text) if PhasePeg.isName(text) => text
        case _ => throw new IllegalArgumentException("rule name required")
      }
      if (rules.contains(name)) { throw new IllegalArgumentException("duplicate rule") }
      expect("=")
      val expr = choice()
      expect(";")
      rules = rules.updated(name, expr)
    }
    rules
  }

  private def peek(): Option[Token] = if (cursor < tokens.length) { Some(tokens(cursor)) } else { None }

  private def peekIs(text: String): Boolean = peek().contains(Punct(text))

  private def take(): Token = {
    val result = peek().getOrElse(throw new IllegalArgumentException("unexpected end of grammar"))
    cursor += 1
    result
  }

  private def expect(text: String): Unit = {
    if (take() != Punct(text)) { throw new IllegalArgumentException(s"expected $text") }
  }

  private def choice(): PegAst = {
    var result = sequence()
    while (peekIs("/")) {
      take()
      result = Choice(result, sequence())
    }
    result
  }

  private def sequence(): PegAst = {
    val items = mutable.ArrayBuffer.empty[PegAst]
    while (peek().isDefined && !peekIs("/") && !peekIs(")") && !peekIs(";")) {
      items += prefix()
    }
    if (items.isEmpty) { throw new IllegalArgumentException("use \"\" for an empty expression") }
    items.tail.foldLeft(items.head)((result, expr) => Sequence(result, expr))
  }

  private def prefix(): PegAst = {
    if (peekIs("!")) {
      take()
      NotPredicate(prefix())
    } else if (peekIs("&")) {
      take()
      AndPredicate(prefix())
    } else {
      var result = primary()
      while (peekIs("*")) {
        take()
        result = Star(result)
      }
      result
    }
  }

  private def primary(): PegAst = {
    take() match {
      case Punct("(") =>
        val result = choice()
        expect(")")
        result
      case Punct(".") => AnyChar
      case Literal(value) =>
        value.foldLeft(PhasePeg.EMPTY) { (result, c) =>
          val char = Terminal(c)
          if (result == PhasePeg.EMPTY) { char } else { Sequence(result, char) }
        }
      case Punct(text) if PhasePeg.isName(text) => Ref(text)
      case Punct(text) => throw new IllegalArgumentException(s"unexpected $text")
    }
  }
}

/** One packrat evaluation of the start rule over `word`.
  *
  * Explicit continuations avoid a host recursion cap on a valid generated
  * grammar. Memo keys use AST identity; rule bodies themselves are shared.
  */
private final class PackratRun(grammar: Grammar, word: String, memo: Memo) {
  import PegAst.*

  private sealed trait Frame
  private final case class Eval(expr: PegAst, position: Int) extends Frame
  private final case class Save(identity: Int, position: Int) extends Frame
  private final case class SeqK(second: PegAst) extends Frame
  private final case class ChoiceK(second: PegAst, position: Int) extends Frame
  private final case class PredicateK(positive: Boolean, position: Int) extends Frame
  private final case class RepeatK(inner: PegAst, position: Int) extends Frame

  def run(): Option[Int] = {
    val work = mutable.Stack[Frame](Eval(grammar.rules(grammar.start), 0))
    var out: Option[Int] = None
    while (work.nonEmpty) {
      work.pop() match {
        case Save(identity, position) => memo.update(identity, position, out.map(_ + 1).getOrElse(0))
        case SeqK(second) => out.foreach(end => work.push(Eval(second, end)))
        case ChoiceK(second, position) => if (out.isEmpty) { work.push(Eval(second, position)) }
        case PredicateK(positive, position) => out = if (out.isDefined == positive) { Some(position) } else { None }
        case RepeatK(inner, position) =>
          out match {
            case None => out = Some(position)
            case Some(end) if end == position => throw new IllegalArgumentException("nullable repetition")
            case Some(end) =>
              work.push(RepeatK(inner, end))
              work.push(Eval(inner, end))
          }
        case Eval(expr, position) =>
          val cached = memo.get(expr.identity, position)
          if (cached >= 0) {
            out = if (cached == 0) { None } else { Some(cached - 1) }
          } else {
            if (cached == -2) { throw new IllegalArgumentException("non-consuming recursion") }
            memo.update(expr.identity, position, -2)
            work.push(Save(expr.identity, position))
            out = evaluate(expr, position, work, out)
          }
      }
    }
    out
  }

  private def evaluate(expr: PegAst, position: Int, work: mutable.Stack[Frame], out: Option[Int]): Option[Int] = {
    expr match {
      case Empty => Some(position)
      case AnyChar => if (position < word.length) { Some(position + 1) } else { None }
      case Terminal(c) => if (position < word.length && word.charAt(position) == c) { Some(position + 1) } else { None }
      case Ref(name) => work.push(Eval(grammar.rules(name), position)); out
      case Sequence(first, second) => work.push(SeqK(second)); work.push(Eval(first, position)); out
      case Choice(first, second) => work.push(ChoiceK(second, position)); work.push(Eval(first, position)); out
      case NotPredicate(inner) => work.push(PredicateK(positive = false, position)); work.push(Eval(inner, position)); out
      case AndPredicate(inner) => work.push(PredicateK(positive = true, position)); work.push(Eval(inner, position)); out
      case Star(inner) => work.push(RepeatK(inner, position)); work.push(Eval(inner, position)); out
    }
  }
}

File: docs/palindromes-in-peg/phase_peg.py
"""Experimental inverse expansion for the plain PEG subset emitted by tm2peg.

inverse_repeat(G, k) recognizes {w | G recognizes each character of w k times}.
The virtual offset inside a repeated character is a finite grammar parameter,
not runtime state. E_i_j succeeds exactly when E returns in phase j from i.
An ordered-choice fallback is guarded by failure of E in *every* return phase.

This is an output-boundary construction, not a real-time scheduling proof.
Input is BMP scalar text. The parser deliberately supports only double-quoted
literals, '.', names, sequence, choice, predicates, grouping and '*'. Source
grammars must terminate (in particular no nullable repetition/left recursion).
"""
import json
import re
from array import array


EMPTY = ("empty",)
TOKEN = re.compile(r"[A-Za-z_][A-Za-z_0-9]*|[=;()/!&.*]")


class CompactMemo:
  """Sparse rows become integer arrays when packrat cells grow dense."""
  def __init__(self, length):
    self.rows, self.width = {}, length + 1
    self.kind, self.missing = ("H", 65535) if length < 65533 else ("Q", 2**64 - 1)
    self.threshold = max(8, self.width // (32 if self.kind == "H" else 8))

  def get(self, key, default=-1):
    identity, position = key
    row = self.rows.get(identity)
    if row is None: return default
    value = row.get(position, self.missing) if isinstance(row, dict) else row[position]
    return -1 if value == self.missing else -2 if value == self.missing - 1 else value

  def __setitem__(self, key, value):
    identity, position = key
    row = self.rows.get(identity)
    if row is None:
      row = self.rows[identity] = {}
    row[position] = self.missing + value + 1 if value < 0 else value
    if isinstance(row, dict) and len(row) > self.threshold:
      dense = array(self.kind, [self.missing]) * self.width
      for index, stored in row.items(): dense[index] = stored
      self.rows[identity] = dense


class Grammar:
  def __init__(self, source, start="S"):
    self.tokens = []
    cursor = 0
    decoder = json.JSONDecoder()
    while cursor < len(source):
      if source[cursor].isspace():
        cursor += 1
      elif source[cursor] == '"':
        value, end = decoder.raw_decode(source, cursor)
        if any(ord(c) > 0xffff or 0xd800 <= ord(c) <= 0xdfff for c in value):
          raise ValueError("BMP scalar literals required")
        self.tokens.append(("literal", value))
        cursor = end
      else:
        found = TOKEN.match(source, cursor)
        if not found:
          raise ValueError(f"unsupported PEG syntax at {cursor}")
        self.tokens.append(found[0])
        cursor += len(found[0])
    self.cursor, self.rules = 0, {}
    while self.cursor < len(self.tokens):
      name = self.take()
      if not isinstance(name, str) or not re.fullmatch(r"[A-Za-z_][A-Za-z_0-9]*", name):
        raise ValueError("rule name required")
      if name in self.rules:
        raise ValueError("duplicate rule")
      self.expect("=")
      expr = self.choice()
      self.expect(";")
      self.rules[name] = expr
    if not self.rules:
      raise ValueError("empty grammar")
    if start not in self.rules:
      raise ValueError(f"unknown start rule: {start}")
    self.start = start

  def peek(self):
    return self.tokens[self.cursor] if self.cursor < len(self.tokens) else None

  def take(self):
    result = self.peek()
    if result is None:
      raise ValueError("unexpected end of grammar")
    self.cursor += 1
    return result

  def expect(self, token):
    if self.take() != token:
      raise ValueError(f"expected {token}")

  def choice(self):
    result = self.sequence()
    while self.peek() == "/":
      self.take()
      result = ("choice", result, self.sequence())
    return result

  def sequence(self):
    items = []
    while self.peek() is not None and self.peek() not in ("/", ")", ";"):
      items.append(self.prefix())
    if not items:
      raise ValueError('use "" for an empty expression')
    result = items[0]
    for expr in items[1:]:
      result = ("seq", result, expr)
    return result

  def prefix(self):
    if self.peek() in ("!", "&"):
      return (self.take(), self.prefix())
    token = self.take()
    if token == "(":
      result = self.choice()
      self.expect(")")
    elif token == ".":
      result = ("any",)
    elif isinstance(token, tuple):
      result = EMPTY
      for c in token[1]:
        char = ("char", c)
        result = char if result == EMPTY else ("seq", result, char)
    elif re.fullmatch(r"[A-Za-z_][A-Za-z_0-9]*", token):
      result = ("ref", token)
    else:
      raise ValueError(f"unexpected {token}")
    while self.peek() == "*":
      self.take()
      result = ("star", result)
    return result

  def accepts(self, word, compact=False):
    return self.parse_prefix(word, compact) == len(word)

  def parse_prefix(self, word, compact=False):
    """Return the start rule's end position, or None on failure.

    Pointer-like PEG rules deliberately leave a suffix. Exposing their result
    lets construction experiments check that suffix, rather than only checking
    whole-word acceptance.
    """
    if any(ord(c) > 0xffff or 0xd800 <= ord(c) <= 0xdfff for c in word):
      raise ValueError("BMP scalar input required")
    memo = CompactMemo(len(word)) if compact else {}
    # Explicit continuations avoid a host recursion cap on a valid generated
    # grammar. Memo keys use AST identity; rule bodies themselves are shared.
    work = [("eval", self.rules[self.start], 0)]
    out = None
    while work:
      task = work.pop()
      action = task[0]
      if action == "save":
        memo[task[1]] = 0 if out is None else out + 1
      elif action == "seq":
        if out is not None: work.append(("eval", task[1], out))
      elif action == "choice":
        if out is None: work.append(("eval", task[1], task[2]))
      elif action == "predicate":
        out = task[2] if (out is not None) == (task[1] == "&") else None
      elif action == "repeat":
        if out is None:
          out = task[2]
        elif out == task[2]:
          raise ValueError("nullable repetition")
        else:
          work.append(("repeat", task[1], out))
          work.append(("eval", task[1], out))
      else:
        expr, pos = task[1], task[2]
        key = (id(expr), pos)
        cached = memo.get(key, -1)
        if cached >= 0:
          out = None if cached == 0 else cached - 1
          continue
        if cached == -2:
          raise ValueError("non-consuming recursion")
        memo[key] = -2
        work.append(("save", key))
        kind = expr[0]
        if kind == "empty":
          out = pos
        elif kind in ("char", "any"):
          out = pos + 1 if pos < len(word) and (kind == "any" or word[pos] == expr[1]) else None
        elif kind == "ref":
          work.append(("eval", self.rules[expr[1]], pos))
        elif kind in ("seq", "choice"):
          work.append((kind, expr[2], pos))
          work.append(("eval", expr[1], pos))
        elif kind in ("!", "&"):
          work.append(("predicate", kind, pos))
          work.append(("eval", expr[1], pos))
        elif kind == "star":
          work.append(("repeat", expr[1], pos))
          work.append(("eval", expr[1], pos))
        else:
          raise ValueError(kind)
    return out



def inverse_repeat(source, width, start="S"):
  if type(width) is not int or width < 1:
    raise ValueError("positive integer width required")
  grammar = Grammar(source, start)
  nodes, ids = [], {}

  def register(expr):
    if expr not in ids:
      ids[expr] = len(nodes)
      nodes.append(expr)
      if expr[0] == "ref":
        register(grammar.rules[expr[1]])
      elif expr[0] in ("seq", "choice", "!", "&", "star"):
        for child in expr[1:]:
          register(child)

  register(grammar.rules[grammar.start])
  # Prune impossible phase returns before rendering. Besides reducing size,
  # this keeps a failed E_i_i from masquerading as a nullable prefix in the
  # target grammar's conservative left-recursion check.
  returns = {expr: set() for expr in nodes}
  diagonal = {(i, i) for i in range(width)}
  changed = True
  while changed:
    changed = False
    for expr in nodes:
      kind = expr[0]
      if kind in ("empty", "!"):
        possible = diagonal
      elif kind in ("char", "any"):
        possible = {(i, (i + 1) % width) for i in range(width)}
      elif kind == "ref":
        possible = returns[grammar.rules[expr[1]]]
      elif kind == "choice":
        possible = returns[expr[1]] | returns[expr[2]]
      elif kind == "seq":
        possible = {(i, j) for i, k in returns[expr[1]]
                    for middle, j in returns[expr[2]] if middle == k}
      elif kind == "&":
        possible = {(i, i) for i, _ in returns[expr[1]]}
      elif kind == "star":
        possible = diagonal | {(i, j) for i, k in returns[expr[1]]
                               for middle, j in returns[expr] if middle == k}
      if not possible <= returns[expr]:
        returns[expr] |= possible
        changed = True
  name = lambda expr, i, j: f"E_{ids[expr]}_{i}_{j}"
  success = lambda expr, i: f"Y_{ids[expr]}_{i}"
  fail = '!""'
  lines = [f"S = {name(grammar.rules[grammar.start], 0, 0)} !.;"]
  literal = lambda c: '"\\u0008"' if c == "\b" else json.dumps(c, ensure_ascii=False)
  for expr in nodes:
    kind = expr[0]
    for i in range(width):
      choices = [f"&{name(expr, i, j)}" for j in range(width) if (i, j) in returns[expr]]
      lines.append(success(expr, i) + " = " + (" / ".join(choices) or fail) + ";")
      for j in range(width):
        alternatives = []
        if (i, j) not in returns[expr]:
          pass
        elif kind == "empty" and i == j:
          alternatives = ['""']
        elif kind in ("char", "any") and j == (i + 1) % width:
          terminal = literal(expr[1]) if kind == "char" else "."
          alternatives = [terminal if i == width - 1 else f"&({terminal})"]
        elif kind == "ref":
          alternatives = [name(grammar.rules[expr[1]], i, j)]
        elif kind == "seq":
          alternatives = [name(expr[1], i, k) + " " + name(expr[2], k, j)
                          for k in range(width)
                          if (i, k) in returns[expr[1]] and (k, j) in returns[expr[2]]]
        elif kind == "choice":
          if (i, j) in returns[expr[1]]:
            alternatives.append(name(expr[1], i, j))
          if (i, j) in returns[expr[2]]:
            other_phase = any(a == i and b != j for a, b in returns[expr[1]])
            commitment = "!" + success(expr[1], i) + " " if other_phase else ""
            alternatives.append(commitment + name(expr[2], i, j))
        elif kind in ("!", "&") and i == j:
          alternatives = [kind + success(expr[1], i)]
        elif kind == "star":
          alternatives = [name(expr[1], i, k) + " " + name(expr, k, j)
                          for k in range(width)
                          if (i, k) in returns[expr[1]] and (k, j) in returns[expr]]
          if i == j:
            alternatives.append("!" + success(expr[1], i))
        lines.append(name(expr, i, j) + " = " + (" / ".join(alternatives) or fail) + ";")
  return "\n".join(lines) + "\n"


if __name__ == "__main__":
  import argparse
  from pathlib import Path
  parser = argparse.ArgumentParser(description=__doc__)
  parser.add_argument("source", type=Path)
  parser.add_argument("width", type=int)
  parser.add_argument("output", type=Path)
  parser.add_argument("--start", default="S")
  args = parser.parse_args()
  args.output.write_text(inverse_repeat(args.source.read_text(), args.width, args.start))
